runtime/lua/vim/lsp/handlers.lua

```diff
@@ -110,36 +110,45 @@ M['client/registerCapability'] = function(_, _, ctx)
 end
 
 --see: https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_codeAction
-M['textDocument/codeAction'] = function(_, result)
-  if result == nil or vim.tbl_isempty(result) then
+M['textDocument/codeAction'] = function(_, actions, ctx)
+  if actions == nil or vim.tbl_isempty(actions) then
     print("No code actions available")
     return
   end
 
   local option_strings = {"Code actions:"}
-  for i, action in ipairs(result) do
+  for i, action in ipairs(actions) do
     local title = action.title:gsub('\r\n', '\\r\\n')
     title = title:gsub('\n', '\\n')
     table.insert(option_strings, string.format("%d. %s", i, title))
   end
 
   local choice = vim.fn.inputlist(option_strings)
-  if choice < 1 or choice > #result then
+  if choice < 1 or choice > #actions then
     return
   end
-  local action_chosen = result[choice]
-  -- textDocument/codeAction can return either Command[] or CodeAction[].
-  -- If it is a CodeAction, it can have either an edit, a command or both.
-  -- Edits should be executed first
-  if action_chosen.edit or type(action_chosen.command) == "table" then
-    if action_chosen.edit then
-      util.apply_workspace_edit(action_chosen.edit)
-    end
-    if type(action_chosen.command) == "table" then
-      buf.execute_command(action_chosen.command)
-    end
+  local action = actions[choice]
+  -- textDocument/codeAction can return either Command[] or CodeAction[]
+  --
+  -- CodeAction
+  --  ...
+  --  edit?: WorkspaceEdit    -- <- must be applied before command
+  --  command?: Command
+  --
+  -- Command:
+  --  title: string
+  --  command: string
+  --  arguments?: any[]
+  --
+  if action.edit then
+    util.apply_workspace_edit(action.edit)
+  end
+  local command = type(action.command) == 'table' and action.command or action
+  local fn = vim.lsp.commands[command.command]
+  if fn then
+    fn(command, ctx)
   else
-    buf.execute_command(action_chosen)
+    buf.execute_command(command)
```

Comment:
I wonder if we should change `buf.execute_command` to do the `vim.lsp.commands` lookup, to ensure all lsp actions which deal with commands check the registry?

(We should also adapt it to use the right client, the `execute_command` here should not go through *all* clients if multiple are attached to the buffer).

Reply:
I think I'm good with that, it makes sense. But is it the case a user might want to register special commands per client?

Reply:
I think I've changed my mind a bit and I'd actually like to defer changing this to a later PR (where I either introduce `codeAction/resolve` support or where I integrate the commands with code-lenses.

In both cases it is necessary to use the right client to use the server side commands, so that will have an influence on the API